src/Voron/Data/CompactTrees/CompactKey.cs

```diff
@@ -0,0 +1,432 @@
+        byte* start = _storage.Ptr + _decodedKeyIdx;
+        int length = *((int*)start);
+
+        return new ReadOnlySpan<byte>(start + sizeof(int), length);
+    }
+
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    public byte* DecodedPtr(out int lengthInBytes)
+    {
+        if (_decodedKeyIdx == Invalid)
+        {
+            DecodeFromEncodedForm();
+        }
+
+        // IMPORTANT: Pointers are potentially invalidated by the grow storage call at DecodeFromEncodedForm, be careful here. 
+        byte* start = _storage.Ptr + _decodedKeyIdx;
+        // This is decoded therefore is bytes.
+        lengthInBytes = *((int*)start);
+
+        return start + sizeof(int);
+    }
+
+    private void UnlikelyGrowStorage(int maxSize)
+    {
+        int memoryUsed = (int)(_currentPtr - _storage.Ptr);
+
+        // Request more memory, copy the content and return it.
+        maxSize = Math.Max(maxSize, _storage.Length) * 2;
+        var storageScope = Owner.Llt.Allocator.Allocate(maxSize, out var storage);
+        Memory.Copy(storage.Ptr, _storage.Ptr, memoryUsed);
+
+        _storageScope.Dispose();
+
+        // Update the new references.
+        _storage = storage;
+        _storageScope = storageScope;
+
+        // This procedure will invalidate any pointer beyond this point. 
+        _currentPtr = _storage.Ptr + memoryUsed;
+        _currentEndPtr = _currentPtr + _storage.Length;
+    }
+
+    public void Set(CompactKey key)
+    {
+        Dictionary = key.Dictionary;
+        _currentKeyIdx = key._currentKeyIdx;
+        _decodedKeyIdx = key._decodedKeyIdx;
+        _lastKeyMappingItem = key._lastKeyMappingItem;
+
+        var originalSize = (int)(key._currentPtr - key._storage.Ptr);
+        if (originalSize > _storage.Length)
+            UnlikelyGrowStorage(originalSize);
+
+        // Copy the key mapping and content.
+        int lastElementIdx = Math.Min(_lastKeyMappingItem, MappingTableSize - 1);
+        if (lastElementIdx >= 0)
+        {
+            var srcDictionary = key._keyMappingCache;
+            var destDictionary = _keyMappingCache;
+            var srcIndex = key._keyMappingCacheIndex;
+            var destIndex = _keyMappingCacheIndex;
+
+            // PERF: Since we are avoiding the cost of general purpose copying, if we have the vector instruction set we should use it. 
+            if (Avx.IsSupported)
+            {
+                int currentElementIdx = 0;
+                while (currentElementIdx <= lastElementIdx)
+                {
+                    Avx.Store(_keyMappingCache + currentElementIdx, Avx.LoadDquVector256(key._keyMappingCache + currentElementIdx));
+                    Avx.Store(_keyMappingCacheIndex + currentElementIdx, Avx.LoadDquVector256(key._keyMappingCacheIndex + currentElementIdx));
+                    currentElementIdx += Vector256<long>.Count;
```

Comment:
Isn't this possible to copy _more_ than the buffer size?

Reply:
Good catch!!! I cannot think how would you end up in that situation 'realistically' but we need to handle that case either way.